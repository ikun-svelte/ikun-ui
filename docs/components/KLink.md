---
title: KLink
lang: en-US
---

# KLink

Text hyperlink

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/link
```

```bash [yarn]
yarn add @ikun-ui/link
```

```bash [npm]
npm install @ikun-ui/link
```

:::

## Basic usage

Basic text link

<demo src="../../../../example/link/basic.svelte" github='link'></demo>

## Disabled

Disabled state of link

<demo src="../../../../example/link/disabled.svelte" github='link'></demo>

## Underline

Underline of link

<demo src="../../../../example/link/underline.svelte" github='link'></demo>

## Icon

Link with icon

<demo src="../../../../example/link/icon.svelte" github='link'></demo>

## link Props

| Name      | Type                                                             | Default   | Description                         |
| --------- | ---------------------------------------------------------------- | --------- | ----------------------------------- |
| type      | `primary` / `success` / `error` / `warning` / `info` / `default` | `default` | Button type                         |
| underline | `boolean`                                                        | `true`    | whether the component has underline |
| disabled  | `boolean`                                                        | `false`   | whether the component is disabled   |
| href      | `string`                                                         | `-`       | same as native hyperlink's href     |
| icon      | `string`                                                         | `-`       | link's icon                         |
| cls       | `string`                                                         | `-`       | Additional class                    |
| attrs     | `Record<string, string>`                                         | `{}`      | Additional attributes               |

## link Slots

| Name    | Description               |
| ------- | ------------------------- |
| default | Customize default content |
| icon    | Customize icon content    |
