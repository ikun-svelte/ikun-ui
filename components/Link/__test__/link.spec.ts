import { afterEach, expect, test, describe, beforeEach, vi } from 'vitest';
import KLink from '../src';
import KLinkIcon from './fixture/link.icon.svelte';
import { tick } from 'svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KLink', () => {
	test('props: cls', async () => {
		const instance = new KLink({
			target: host,
			props: {
				cls: 'k-link--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-link--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: type', async () => {
		const instanceDefault = new KLink({
			target: host
		});
		expect(instanceDefault).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-link--default k-link--default__hover')
		).toBeTruthy();

		const instanceInfo = new KLink({
			target: host,
			props: {
				type: 'info'
			}
		});
		expect(instanceInfo).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-link--info k-link--info__hover')
		).toBeTruthy();

		const instanceError = new KLink({
			target: host,
			props: {
				type: 'error'
			}
		});
		expect(instanceError).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-link--error k-link--error__hover')
		).toBeTruthy();

		const instanceWarning = new KLink({
			target: host,
			props: {
				type: 'warning'
			}
		});
		expect(instanceWarning).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-link--warning k-link--warning__hover')
		).toBeTruthy();

		const instanceSuccess = new KLink({
			target: host,
			props: {
				type: 'success'
			}
		});
		expect(instanceSuccess).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-link--success k-link--success__hover')
		).toBeTruthy();

		const instancePrimary = new KLink({
			target: host,
			props: {
				type: 'primary'
			}
		});
		expect(instancePrimary).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-link--primary k-link--primary__hover')
		).toBeTruthy();
	});

	test('props: href', async () => {
		const instance = new KLink({
			target: host,
			props: {
				href: 'jinitaimei'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('href="jinitaimei"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: icon', async () => {
		const instance = new KLink({
			target: host,
			props: {
				icon: 'i-carbon-logo-vue'
			}
		});
		expect(instance).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-icon--base k-icon-transition i-carbon-logo-vue')
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: underline', async () => {
		const instance = new KLink({
			target: host,
			props: {
				underline: false
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-link--underline__none')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		const instance = new KLink({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		expect(
			(host as HTMLElement)!.innerHTML.includes('k-link--disabled k-link--underline__none')
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('events: click', async () => {
		const mockFn = vi.fn();
		const instance = new KLink({
			target: host
		});
		await tick();
		instance.$on('click', mockFn);
		expect(instance).toBeTruthy();
		const btn = host.getElementsByTagName('a')[0];
		btn.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(mockFn).toBeCalled();
	});

	test('events: disabled & click', async () => {
		const mockFn = vi.fn();
		const instance = new KLink({
			target: host,
			props: {
				disabled: true
			}
		});
		await tick();
		instance.$on('click', mockFn);
		expect(instance).toBeTruthy();
		const btn = host.getElementsByTagName('a')[0];
		btn.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(mockFn).not.toBeCalled();
	});

	test('slots: icon', async () => {
		const instance = new KLinkIcon({
			target: host
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('i-carbon-logo-vue')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
