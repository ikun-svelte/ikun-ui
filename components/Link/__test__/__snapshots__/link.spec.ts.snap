// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KLink > props: cls 1`] = `"<a class=\\"k-link k-link--base k-link--default k-link--default__hover k-link--underline k-link--test\\" aria-hidden=\\"true\\">  </a>"`;

exports[`Test: KLink > props: disabled 1`] = `"<a class=\\"k-link k-link--base k-link--default k-cur-disabled k-link--disabled k-link--underline__none\\" aria-hidden=\\"true\\">  </a>"`;

exports[`Test: KLink > props: href 1`] = `"<a class=\\"k-link k-link--base k-link--default k-link--default__hover k-link--underline\\" aria-hidden=\\"true\\" href=\\"jinitaimei\\">  </a>"`;

exports[`Test: KLink > props: icon 1`] = `"<a class=\\"k-link k-link--base k-link--default k-link--default__hover k-link--underline\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-logo-vue ikun-initial\\" role=\\"\\" aria-hidden=\\"true\\"></span>  </a>"`;

exports[`Test: KLink > props: underline 1`] = `"<a class=\\"k-link k-link--base k-link--default k-link--default__hover k-link--underline__none\\" aria-hidden=\\"true\\">  </a>"`;

exports[`Test: KLink > slots: icon 1`] = `"<a class=\\"k-link k-link--base k-link--success k-link--success__hover k-link--underline\\" target=\\"_blank\\" aria-hidden=\\"true\\">  <div class=\\"i-carbon-logo-vue\\" slot=\\"icon\\"></div></a>"`;
